File: typescript/examples/string-truncate.test.ts
// カード string-truncate の Contract を検証するテスト
import { truncate } from 'es-toolkit/compat';
import { describe, expect, it } from 'vitest';

describe('string-truncate: truncate', () => {
  const text = 'The quick brown fox jumps over the lazy dog';

  it('length は省略記号を含めた最大長で、既定は 30 と "..."', () => {
    expect(truncate(text)).toBe('The quick brown fox jumps o...');
    expect(truncate(text)).toHaveLength(30);
    expect(truncate('abcdefghijkl', { length: 10 })).toBe('abcdefg...');
  });

  it('length 以下の文字列は変更しない', () => {
    expect(truncate('abc', { length: 3 })).toBe('abc');
    expect(truncate('abc', { length: 10 })).toBe('abc');
  });

  it('文字列が省略記号以下の長さ、または length が省略記号より短ければ省略記号だけを返す', () => {
    expect(truncate('ABC', { length: 2 })).toBe('...');
    expect(truncate('A', { length: 2 })).toBe('A'); // 切り詰め不要なら省略記号だけにはならない
    expect(truncate('abcdef', { length: 2, omission: '[...]' })).toBe('[...]');
  });

  it('separator を指定すると手前にある最後の区切りで切り、無ければそのまま切る', () => {
    expect(truncate(text, { length: 20, separator: ' ' })).toBe('The quick brown...');
    expect(truncate(text, { length: 20, separator: /,? +/ })).toBe('The quick brown...');
    expect(truncate('abcdefghijkl', { length: 8, separator: ' ' })).toBe('abcde...');
  });

  it('omission を差し替えられる', () => {
    expect(truncate(text, { length: 20, omission: '…' })).toBe('The quick brown fox…');
    expect(truncate('abcdefghijkl', { length: 8, omission: '' })).toBe('abcdefgh');
  });

  it('サロゲートペアを含む文字列はコードポイント単位で数える', () => {
    expect(truncate('¥§✈✉🤓', { length: 5 })).toBe('¥§✈✉🤓');
    expect(truncate('😀😀😀😀😀', { length: 4, omission: '…' })).toBe('😀😀😀…');
    expect(truncate('あいうえおかきくけこ', { length: 6 })).toBe('あいう...');
  });

  it('length が 0 以下なら 0 扱い、undefined は空文字を返し、例外を投げない', () => {
    expect(truncate('ABC', { length: 0 })).toBe('...');
    expect(truncate('ABC', { length: -5 })).toBe('...');
    expect(truncate(undefined)).toBe('');
    expect(truncate('')).toBe('');
  });

  it('同じ入力には同じ結果を返す', () => {
    expect(truncate(text, { length: 12 })).toBe(truncate(text, { length: 12 }));
  });
});

File: go/examples/string-truncate_test.go
// カード string-truncate の Contract を検証するテスト
package examples

import (
	"testing"
	"unicode/utf8"

	"github.com/samber/lo"
)

func TestStringTruncate(t *testing.T) {
	text := "The quick brown fox jumps over the lazy dog"

	t.Run("length は省略記号を含めた最大ルーン数", func(t *testing.T) {
		if got := lo.Ellipsis(text, 30); got != "The quick brown fox jumps o..." {
			t.Errorf("Ellipsis(30) = %q", got)
		}
		if got := lo.Ellipsis(text, 20); got != "The quick brown f..." {
			t.Errorf("Ellipsis(20) = %q", got)
		}
		if got := utf8.RuneCountInString(lo.Ellipsis(text, 20)); got != 20 {
			t.Errorf("ルーン数 = %d", got)
		}
	})

	t.Run("ルーン単位で数え、バイト長ではない", func(t *testing.T) {
		got := lo.Ellipsis("日本語の文章です", 5)
		if got != "日本..." {
			t.Errorf("日本語 = %q", got)
		}
		if utf8.RuneCountInString(got) != 5 || len(got) != 9 {
			t.Errorf("runes=%d bytes=%d", utf8.RuneCountInString(got), len(got))
		}
		if got := lo.Ellipsis("🐶🐶🐶🐶🐶", 4); got != "🐶..." {
			t.Errorf("絵文字 = %q", got)
		}
		// ZWJ で結合した絵文字はコードポイントごとに分かれる
		family := "👨‍👩‍👧"
		if got := lo.Ellipsis(family+" family", 6); got != "👨‍👩..." {
			t.Errorf("ZWJ = %q", got)
		}
	})

	t.Run("前後の空白と切り位置の末尾の空白を落とす", func(t *testing.T) {
		if got := lo.Ellipsis("  padded text  ", 20); got != "padded text" {
			t.Errorf("TrimSpace = %q", got)
		}
		if got := lo.Ellipsis("hello world", 9); got != "hello..." {
			t.Errorf("末尾空白 = %q", got)
		}
	})

	t.Run("ルーン数が length 以下ならそのまま返す", func(t *testing.T) {
		if got := lo.Ellipsis(text, 43); got != text {
			t.Errorf("ちょうど = %q", got)
		}
		if got := lo.Ellipsis(text, 100); got != text {
			t.Errorf("余裕 = %q", got)
		}
		if got := lo.Ellipsis("ab", 2); got != "ab" {
			t.Errorf("短い = %q", got)
		}
	})

	t.Run("length が 3 以下で切り詰めが必要なら ... だけを返し length を超える", func(t *testing.T) {
		for _, l := range []int{3, 2, 1, 0, -1} {
			if got := lo.Ellipsis("abcd", l); got != "..." {
				t.Errorf("Ellipsis(abcd, %d) = %q", l, got)
			}
		}
		if got := lo.Ellipsis("abc", 3); got != "abc" {
			t.Errorf("Ellipsis(abc, 3) = %q", got)
		}
	})

	t.Run("空文字は空文字", func(t *testing.T) {
		if got := lo.Ellipsis("", 0); got != "" {
			t.Errorf("Ellipsis(\"\", 0) = %q", got)
		}
		if got := lo.Ellipsis("", 5); got != "" {
			t.Errorf("Ellipsis(\"\", 5) = %q", got)
		}
	})
}
